### PR TITLE
Fixing according to CR

### PR DESCRIPTION
**Pull request by eladiw**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,72 +1,86 @@
 
+# Helm Charts installation through http endpoints
 
-# Project Autom8s - Automation for K8S
```
Comment (limorl) [inline README.md:3]: I am not sure 'automation for K8s reflects what this repo does, you should consider another title and repo name

File: README.md [hunk 2]
```diff
@@ -1,72 +1,86 @@
...
+    ```
+
+Now you have a working Autom8s API awaiting HTTP requests.
 
 ## Using the API
+
 If you used the default settings, the API will be accessible internally at: `http://autom8s.default.svc.cluster.local:4000`
 
-Here is a quick node.js snippet that makes use of the API to install RabbitMQ with default settings:
+Here is a quick node.js snippet that makes use of the API to install RabbitMQ with default settings. Note: any chart will be suitable, RabbitMQ is just a specific example:
```
Comment (limorl) [inline README.md]: 'Note: ...' in a new line

File: README.md [hunk 3]
```diff
@@ -1,72 +1,86 @@
...
 let chart = { name: "stable/rabbitmq", servicePort: 5672 };
 
 // perform helm install
-var installResponse = await requestPostAsync(Paths.HelmInstall, { form: { chartName: chart.name } });
+var installResponse = await requestPostAsync(Paths.HelmInstall, { form: { chartName: "" } });
 
 // create a rule to expose the new service expternally
 var ingressResponse = await requestGetAsync(Paths.SetIngressRule, { serviceName: installResponse.serviceName, servicePort: chart.servicePort });
 
 return "Your new service: " + ingressResponse.releaseName + ", is publicly accessibly on " + ingressResponse.ip + ":" + ingressResponse.port;
```
Comment (limorl) [inline README.md]: publicly accessible (instead of accessibly)

Comment (limorl) [inline README.md]: please use format

File: autom8s/helm-controller.js [hunk 4]
```diff
@@ -25,6 +27,9 @@ router.post('/install',
       });
   });
 
+/**
+ * Deletes an already installed chart, identified by its release name
+ */
```
Comment (limorl) [inline autom8s/helm-controller.js]: Here and everywhere else: when writing a service and returning an HTTP error, don't expose the internal error and avoid returning it. It is not a good practice.
 You should log the error, and if you have a user understandable error message, you may return it.

File: autom8s/helm.js [hunk 5]
```diff
@@ -10,22 +10,26 @@ const helmBinaryLocation = process.env.HELM_BINARY;
...
+    console.log(`Installing new chart. deployOptions: ${JSON.stringify(deployOptions)}`);
     const chartName = deployOptions.chartName.toLowerCase();
     const { releaseName } = deployOptions;
     let installCommand = `json install ${chartName}`;
 
+    // sanity
+    Helm._verifyNotEmpty(chartName, 'chartName');
+
     if (releaseName !== undefined && releaseName != null && releaseName !== '') {
+      console.log(`using a user chosen release name: ${releaseName}`);
```
Comment (limorl) [inline autom8s/helm.js]: change log to: installing specified release name: ${releaseName}

File: autom8s/helm.js [hunk 6]
```diff
@@ -10,22 +10,26 @@ const helmBinaryLocation = process.env.HELM_BINARY;
...
       installCommand = `${installCommand} --name ${releaseName.toLowerCase()}`;
-      console.log(`Install command: ${installCommand}`);
     }
 
-    return this._innerInstallUpgrade(installCommand, deployOptions)
```
Comment (limorl) [inline autom8s/helm.js:28]: remove the prefix 'inner' from every method. The '_' prefix indicates that

File: autom8s/helm.js [hunk 7]
```diff
@@ -10,22 +10,26 @@ const helmBinaryLocation = process.env.HELM_BINARY;
...
+    console.log(`Install command: ${installCommand}`);
+    return this._installOrUpgradeChart(installCommand, deployOptions)
       .then((responseData) => {
         if (responseData && responseData.error) {
           throw new Error(`Install command failed: ${responseData.error}`);
```
Comment (limorl) [inline autom8s/helm.js]: don't you want to log the error?

File: autom8s/helm.js [hunk 8]
```diff
@@ -48,18 +52,30 @@ class Helm {
...
 
+    console.log(`deleting release: ${releaseName}`);
     return this._executeHelm(`delete ${releaseName}`);
   }
 
   async upgrade(deployOptions) {
     const chartName = deployOptions.chartName.toLowerCase();
     const releaseName = deployOptions.releaseName.toLowerCase();
+
+    Helm._verifyNotEmpty(chartName, 'chartName');
```
Comment (limorl) [inline autom8s/helm.js]: better rename to _validateNotEmpty or _assertNotEmpty - which is more conventional

File: autom8s/helm.js [hunk 9]
```diff
@@ -48,18 +52,30 @@ class Helm {
...
     const upgradeCommand = `upgrade ${releaseName} ${chartName}`;
-    console.log(`upgradeCommand command: ${upgradeCommand}`);
+    console.log(`upgrade command: ${upgradeCommand}`);
+    return this._installOrUpgradeChart(upgradeCommand, deployOptions);
+  }
 
-    return this._innerInstallUpgrade(upgradeCommand, deployOptions);
+  static _verifyNotEmpty(arg, argName) {
+    if (typeof arg === 'undefined' || arg === null || arg === '') {
+      const errorMsg = `${argName} is mandatory`;
```
Comment (limorl) [inline autom8s/helm.js]: change 'mandatory' to 'required' whch is more conventional (e.g. required or optional parameters)

File: autom8s/helm.js [hunk 10]
```diff
@@ -48,18 +52,30 @@ class Helm {
...
+      console.error(errorMsg);
+      throw new Error(errorMsg);
+    }
   }
 
   static _findFirstService(json) {
```
Comment (limorl) [inline autom8s/helm.js:88]: Not clear to me why you are searching for v1/Service. will it always be v1?

File: autom8s/ingress-controller.js [hunk 11]
```diff
@@ -12,7 +15,7 @@ router.get('/setrule',
       specificlb,
```
Comment (limorl) [inline autom8s/ingress-controller.js]: the prefix 'specific' is redundant

File: index.js [hunk 12]
```diff
@@ -1,19 +1,17 @@
...
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 
-app.use('/', helmServer);
-app.use('/', portServer);
-app.use('/', ingressServer);
-app.use('/', sampleApp);
+app.use('/', helmController);
+app.use('/', portController);
+app.use('/', ingressController);
```
Comment (limorl) [inline index.js]: I have told Itye that a better practice would be to have all the HTTP commands here (app.post, app.get, etc.) which makes very clear what is the API. And then you don't have xxx-service, but rather just xxx classes..
app.use hides the interface. 
I don't like it.

File: README.md (hunk 1 above)
Comment (eladiw) [reply README.md:3]: Agree, the new name is : "Helm Charts installation through http endpoints"

File: autom8s/helm.js (hunk 6 above)
Comment (eladiw) [reply autom8s/helm.js:28]: It is already removed (note the red marking on that line)

File: autom8s/helm.js (hunk 10 above)
Comment (eladiw) [reply autom8s/helm.js:88]: Currently yes, v1/service is the only option (up to k8s 1.11). however, i made it more robust by ignoring the version section

**Review by eladiw — COMMENTED** [review]
Pushed new commit with all the fixes

File: index.js [hunk 13]
```diff
@@ -1,17 +1,139 @@
...
 
-app.use('/', helmController);
-app.use('/', portController);
-app.use('/', ingressController);
+// Helm functionallity
+
+/**
+ * Installs the requested chart into the Kubernetes cluster
+ */
+app.post('/install',
```
Comment (limorl) [inline index.js:17]: BTW, you could still keep the logic in the port-controller, ingress-controller objects, etc, and only extract the http service related functionality under app.post, app.get, etc.

My comment was about making sure the api is clear from going over the index.js file without having to go over 4 different files.

**Review by limorl — APPROVED** [review]

File: README.md [hunk 14]
```diff
@@ -1,72 +1,88 @@
...
 
-The solution we propose consists of two services:
+The solution exposes web endpoint which provides the Helm functionalities of Install, Delete and Upgrade charts, as well as automatically configure the ingress controller to expose them to the internet all within a single public IP.
+The solution we propose consists of two parts, both as web servers:
 
-**Helm as a service**: let developers manage helm charts from inside the cluster, using a simple REST API.
+**Helm charts deployer**: let developers install, delete and upgrade helm charts from inside the cluster, using a simple REST API.
 
-**Expose as a service**: expose installed helm charts to the internet, via a single IP.
+**Ingres rule setter**: expose installed helm charts to the internet, via a single IP.
```
Comment (torosent) [inline README.md]: Ingress

File: README.md [hunk 15]
```diff
@@ -1,72 +1,88 @@
...
+* Helm
+
+### Installation
+
+1. Install [Helm](https://github.com/kubernetes/helm) with [RBAC](https://github.com/kubernetes/helm/blob/master/docs/rbac.md#tiller-and-role-based-access-control)
+
+     Make sure to grant tiller sufficient permissions to run helm inside the cluster, and install the Autom8s Chart.
+    The example below will configure helm and tiller to work with the chart's default values. Execute the following lines if using the default chart values:
+    ```bash
+    kubectl apply -f https://raw.githubusercontent.com/itye-msft/kubernetes-dynamic-deployment-service/master/rbac-example/tiller.yaml
```
Comment (torosent) [inline README.md]: this link will change once you transfer repo

File: README.md [hunk 16]
```diff
@@ -75,102 +91,16 @@ Installig a private chart is also simple:
...
+Check out the API documentation [here](./docs/api.md)
 
-For example: http://<automates-service-url>/getport?lbip=1.2.3.4
+## How it works
 
- 
+Feel free to read the extended summary [here](./docs/deepdive.md)
 
-Finding a free port follows a simple pattern:
+## Contributing
```
Comment (torosent) [inline README.md]: This should have an official MS OSS notice

**Review by torosent — CHANGES_REQUESTED** [review]

**Pull request merged**